### PR TITLE
Replace StringBuilder with TextWriter in JsonFormatter

### PR DESCRIPTION
**Pull request by avgweb**
This commit simply replace StringBuilder with TextWriter to avoid string allocation/encoding overhead when dealing with streams.

### CODE REVIEW

**Comment by avgweb** [comment]
I signed it!

**Comment by haberman** [comment]
@jskeet want to take a look?

**Comment by jskeet** [comment]
Yup, happy to.

File: csharp/src/Google.Protobuf/Collections/MapField.cs [hunk 1]
```diff
@@ -474,9 +475,9 @@ public int CalculateSize(Codec codec)
         /// </summary>
         public override string ToString()
         {
-            var builder = new StringBuilder();
-            JsonFormatter.Default.WriteDictionary(builder, this);
-            return builder.ToString();
+            var writer = new StringWriter();
+            JsonFormatter.Default.WriteDictionary(writer, this);
+            return writer.GetStringBuilder().ToString();
```
Comment (jskeet) [inline csharp/src/Google.Protobuf/Collections/MapField.cs]: Just writer.ToString() should be fine here.

File: csharp/src/Google.Protobuf/JsonFormatter.cs [hunk 2]
```diff
@@ -142,16 +142,37 @@ public JsonFormatter(Settings settings)
         public string Format(IMessage message)
         {
             Preconditions.CheckNotNull(message, nameof(message));
-            StringBuilder builder = new StringBuilder();
+            var writer = new StringWriter();
```
Comment (jskeet) [inline csharp/src/Google.Protobuf/JsonFormatter.cs]: I'd prefer this to just delegate to the new method:

```csharp
public string Format(IMessage message)
{
    var writer = new StringWriter();
    Format(message, writer); // Will handle null validation
    return writer.ToString();
}
```

**Comment by jskeet** [comment]
Other than the comments I've left, this looks good. I'm pleased that it only adds to the public API :)
It's possible that we should expose the existing functionality of parsing JSON from a `TextReader` in `IMessageParser` too, but that can come later.

Proposed next steps:
- You (avgweb) add commit with just the changes I've suggested, where you agree with them
- I give an LGTM to that (I would expect, anyway!)
- You rebase to master to avoid the merge conflicts, then rebase interactively to squash the two commits to one
- I merge

Sound okay?

**Comment by avgweb** [comment]
Yep, sounds right. I'll do it later today.

**Comment by avgweb** [comment]
@jskeet Please review

**Comment by jskeet** [comment]
Code looks good now, thanks. It would be good to see this all as just one commit though - in particular, this includes commits that have already happened in the master, which is odd.

If you do

```
$ git rebase master
```

then I _think_ you should end up with just two commits here. At that point, you can use

```
$ git rebase -i HEAD~2
```

and make the second commit a "fixup". Then you should basically just have one commit relative to master.

At that point, force-push into this fork, so we can see just the one commit, and I'll merge it. I don't mind if it ends up as two commits rather than one, but I'd be nervous at seeing the other commits in here.

**Comment by avgweb** [comment]
@jskeet Cannot get rid of the first commit. I'm going to create a new pull request for the changes.

**Pull request closed without merging**